elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/QueryEngine.java

```diff
@@ -0,0 +1,161 @@
+ */
+package com.yahoo.elide.datastores.aggregation;
+
+import com.yahoo.elide.core.DataStore;
+import com.yahoo.elide.core.DataStoreTransaction;
+import com.yahoo.elide.core.RequestScope;
+import com.yahoo.elide.core.filter.expression.FilterExpression;
+import com.yahoo.elide.core.filter.expression.FilterExpressionVisitor;
+import com.yahoo.elide.core.hibernate.Query;
+import com.yahoo.elide.core.pagination.Pagination;
+import com.yahoo.elide.core.sort.Sorting;
+import com.yahoo.elide.datastores.aggregation.filter.expression.HavingFilterExpression;
+import com.yahoo.elide.datastores.aggregation.filter.expression.WhereFilterExpression;
+
+import java.util.Optional;
+
+/**
+ * A {@link QueryEngine} is an abstraction layer that {@link DataStoreTransaction} leverages to optimize query and send
+ * query, such as loading objects, to a persistent storage.
+ * <p>
+ * The purpose of {@link QueryEngine} is to allow a single {@link DataStore} to utilize multiple query frameworks, such
+ * as Hibernate on SQL and Drill on NoSQL shown below:
+ * <pre>
+ *                              +-----------+
+ *                              |           |
+ *                              | DataStore |
+ *                              |           |
+ *                              +-----------+
+ *                                    |
+ *                                    |
+ *                         +----------v-----------+                    +----------------------+
+ *                         |                      |                    |                      |
+ *                         | DataStoreTransaction |                +---+ HibernateQueryEngine |
+ *                         |                      |                |   |                      |
+ *                         +----------------------+                |   +----------------------+
+ *                                    |                            |
+ *                                    |                            |
+ *                             +------v------+                     |   +------------------+
+ *                             |             |                     |   |                  |
+ *                             | QueryEngine +-------------------------+  JpaQueryEngine  |
+ *                             |             |                     |   |                  |
+ *                             +-------------+                     |   +------------------+
+ *                                    |                            |
+ *                                    |                            |
+ *          +--------------------------------------                |   +------------------+
+ *          |                |                    |                |   |                  |
+ * +--------v-------+   +----v----+    +----------v-----------+    +---+ DrillQueryEngine |
+ * |    Hibernate   |   |   JPA   |    |     Apache Drill     |        |                  |
```

Comment:
Drill would still use a SQL driver in this diagram.

Reply:
I've updated the diagram